### PR TITLE
Add the SPI RGB code

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by nomis52** [comment]
Should be ready for review now. This is just basic SPI support for 2 pixels. I'll work on hooking up the RDM personalities / manufacturer PIDs in the next change.

File: firmware/src/app.c [hunk 1]
```diff
@@ -80,14 +81,26 @@ void APP_Initialize(void) {
...
   Logger_Initialize(NULL, PAYLOAD_SIZE);
   Logger_SetState(true);
+
+  SPIRGBConfiguration spi_config;
+  spi_config.module_id = SPI_ID_1;
+  spi_config.baud_rate = 1000000;
+  spi_config.use_enhanced_buffering = true;
+  SPIRGB_Init(&spi_config);
+
+  // Send a frame with all pixels set to 0.
```
Comment (peternewman) [inline firmware/src/app.c]: Is that a good idea/necessary? You'll then get a flash to black if you repower it.

File: firmware/src/responder.c [hunk 2]
```diff
@@ -200,6 +207,13 @@ void Responder_Receive(const TransceiverEvent *event) {
         }
         break;
       case STATE_DMX_DATA:
+        // TODO(simon): configure this with DMX_START_ADDRESS and footprints.
+        if (g_offset - 1 < 6) {
```
Comment (peternewman) [inline firmware/src/responder.c]: The 6 is for two pixels yes? Can we at least make it 3*2 for now?

File: firmware/src/spi_rgb.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ * spi_rgb.c
+ * Copyright (C) 2015 Simon Newton
+ */
+
+#include "spi_rgb.h"
+
+#include "peripheral/spi/plib_spi.h"
+#include "syslog.h"
+
+// TODO(simon): move these into the config (and set with RDM?)
```
Comment (peternewman) [inline firmware/src/spi_rgb.c]: Yes to the RDM bit. We should define a few manufacturer PIDs which are common between this and OLA. I also think we should split the protocol from the personality. Pixel count and protocol seems like the only stuff we'll need to set via RDM, slots per pixel and latch bytes can be inferred from the protocol.

File: firmware/src/spi_rgb.c [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+#define PIXEL_COUNT 2
+#define SLOT_PER_PIXEL 3
```
Comment (peternewman) [inline firmware/src/spi_rgb.c]: SPaG SLOTS_PER_PIXEL

File: firmware/src/spi_rgb.c [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  g_spi.use_enhanced_buffering = config->use_enhanced_buffering;
+  g_spi.in_update = false;
+  g_spi.tx_index = 0;
+
+  unsigned int i = 0;
+  for (; i < SLOT_PER_PIXEL * PIXEL_COUNT; i++) {
+    g_spi.pixels[i] = LPD8806_PIXEL_BYTE;
+  }
+
+  for (i = SLOT_PER_PIXEL * PIXEL_COUNT;
```
Comment (peternewman) [inline firmware/src/spi_rgb.c]: Can we not do memset?

File: firmware/src/responder.c [hunk 6]
```diff
@@ -200,6 +207,13 @@ void Responder_Receive(const TransceiverEvent *event) {
...
+          SPIRGB_SetPixel((g_offset - 1) / 3, (g_offset - 1) % 3, b);
```
Comment (peternewman) [inline firmware/src/responder.c]: Does this need a cast to an int?

File: tests/mocks/SPIRGBMock.cpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * SPIRGBMock.cpp
+ * A mock coarse timer module.
```
Comment (peternewman) [inline tests/mocks/SPIRGBMock.cpp]: Correct

File: tests/mocks/SPIRGBMock.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * SPIRGBMock.h
+ * A mock coarse timer module.
```
Comment (peternewman) [inline tests/mocks/SPIRGBMock.h]: Correct

**Comment by peternewman** [comment]
Generally LGTM, just a few comments

File: firmware/src/app.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/app.c]: Yeah, the SPI bus seems to spew random data on boot, so you get multi-colored rainbows. Resetting to black seems the most sensible thing to do.

File: firmware/src/responder.c (hunk 2 above)
Comment (nomis52) [inline firmware/src/responder.c]: Already fixed in the next change :)

File: firmware/src/spi_rgb.c (hunk 3 above)
Comment (nomis52) [inline firmware/src/spi_rgb.c]: Yeah, I was trying to decide if we do personalities or manufacturer PIDs. The downside of manufacturer PIDs is that we can't yet set them in the OLA web UI. But they are much more powerful that trying to shoehorn personalities, so I think I'll go that direction.

File: firmware/src/spi_rgb.c (hunk 4 above)
Comment (nomis52) [inline firmware/src/spi_rgb.c]: Fixed

File: firmware/src/spi_rgb.c (hunk 5 above)
Comment (nomis52) [inline firmware/src/spi_rgb.c]: fixed

File: firmware/src/responder.c (hunk 6 above)
Comment (nomis52) [inline firmware/src/responder.c]: Nope

File: tests/mocks/SPIRGBMock.cpp (hunk 7 above)
Comment (nomis52) [inline tests/mocks/SPIRGBMock.cpp]: Fixed

File: tests/mocks/SPIRGBMock.h (hunk 8 above)
Comment (nomis52) [inline tests/mocks/SPIRGBMock.h]: Fixed

**Comment by peternewman** [comment]
LGTM

File: firmware/src/app.c (hunk 1 above)
Comment (peternewman) [inline firmware/src/app.c]: Understood.

File: firmware/src/responder.c (hunk 2 above)
Comment (peternewman) [inline firmware/src/responder.c]: Okay

File: firmware/src/spi_rgb.c (hunk 3 above)
Comment (peternewman) [inline firmware/src/spi_rgb.c]: We can easily add manufacturer PIDs to the OLA web UI though. We should try and standardise on a PID across the two codebases anyway.

**Pull request merged**